Updated LeafletOsmMap to remove unSaved markers if a user clicks else…

steveone opened the pull request:
Issue Number: 226

Issue Description: If you click on multiple points of the map, each location will get a pin without you hitting "save" on the pins. The expected behavior is that one pin will be created, and if you don't hit "save" the pin will be removed when you click elsewhere on the map.

Summary of solution:

Added "unSavedMarker" to components local state to track unsaved markers
When a new marker is created, the old marker is removed
Can this issue be closed?

Additional code needed to clear "unSavedMarker" in the state when a marker is successfully saved, that feature is throwing an error (in Chrome Canary) so testing was not possible

Should any new issues be added as a result of this solution?

Saving a marker throws an error
Uncaught (in promise) TypeError: Failed to fetch xhr.js....

jkwening commented on client/src/components/LeafletOsmMap/index.js:
```diff
@@ -171,6 +172,13 @@ export default class LeafletOSMMap extends Component {
+    if (this.state.unSavedMarker) {
+          console.log(this.state.unSavedMarker);
+          this.state.unSavedMarker.remove();
+          dropPin(this.state.UnSavedMarker, event.target);
```
I don't think you need this code snippet - this function is supposed to assist add a pin to the map. Isn't the purpose of this if scope to remove any "unSavedMarker" from the map?

jkwening commented on client/src/components/LeafletOsmMap/index.js:
```diff
@@ -184,6 +192,8 @@ export default class LeafletOSMMap extends Component {
         //remove old icon
         droppedPin.remove()
 
+        //TODO: delete temporary marker from unSavedMarker
```
isn't this currently handled in the `onMapClick()` function? I don't think think we need to implement this.

jkwening commented on client/src/components/LeafletOsmMap/index.js:
```diff
@@ -213,12 +223,13 @@ export default class LeafletOSMMap extends Component {
 
     L.DomEvent.on(deleteBtn, 'click', function() {
       droppedPin.remove();
+      //TODO: delete temporary marker from unSavedMarker
```
Ditto - I don't think needs to be implemented.

steveone commented:
@jkwening 

I have replicated the issue in the code I pulled this morning (see attached). Without my code you get the "Before_new_Code" picture where you can see multiple markers
<img width="281" alt="after_new_code" src="https://user-images.githubusercontent.com/13615356/37550830-35992fac-2951-11e8-929c-7c2f481a7c20.png">

With my code (lines 175-178) you get the "after_new_code" picture with only one marker after 5 or so clicks on the map. 

<img width="289" alt="before_new_code" src="https://user-images.githubusercontent.com/13615356/37550831-35abdf4e-2951-11e8-931e-052affabe7ea.png">

The comment I put in for on line 232 may not need new code since hitting the delete button does remove the marker. The comment at line 195 ( //TODO: delete temporary marker from unSavedMarker) is to write code to clear the unSavedMarker in state. I couldn't write/test the code since the "SaveBtn" wasn't working in the code I had.

Let me know if you have any further questions.

jkwening commented:
@steveone as I noted, the code in this PR resolves the multiple dropped pin issue. What I couldn't replicate is the issue you mention about "SaveBtn" not working. It's working fine on my end - this feature was completed a week or so. I'm able to save a newly dropped pin without any errors. Can you create an issue with the screenshot of the error or stack trace if you'r still getting this error after merging in the current development branch?

Regarding the to statements, if I understand correctly, this not actually a delete but rather setting the unSavedMarker state to null?

steveone commented:
@jkwening I have the latest code as of 3/16 in the morning and get the following errors when I click on the "saveBtn" (it appears to not work since the button doesn't  go away). I am using Chrome Safari on mac OSx

<img width="819" alt="screen shot 2018-03-17 at 9 21 04 am" src="https://user-images.githubusercontent.com/13615356/37557630-a9a5b244-29c4-11e8-8216-4bdf5cf11f6c.png">

You are correct that the "Todo" comments are intended simply to set {"unSavedMarker" : NULL} in state when someone hits the saveBtn.

motosharpley commented:
I can't seem to replicate this one either, are you starting mongodb prior to starting the server?

steveone commented:
@motosharpley yes, I have MongoDB running as well as the server and client.

steveone commented:
I've updated the code in the "saveBtn" to clear the unSavedMarker and removed the //TODO comment from the "deleteBtn" as the code would have been unnecessary.

jkwening commented:
wonderful - going to go ahead and merge this in now. Thanks.
